---
title: ""
author: ""

format:
    docx: default
    typst:
        mainfont: "Ubuntu"
        keep-typ: true

tbl-cap-location: top

execute:
  echo: false
  warning: false

---


Table 1. Summary of principal results generated by the MTBseq pipeline.

| Report name                    | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | Primitive step name | File location in MTBseq results                                                                                                |
|--------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------|--------------------------------------------------------------------------------------------------------------------------------|
| Classification                 | For lineage classification based on a set of phylogenetic SNPs. Within this tab-delimited file, the majority lineage is reported for each dataset. This entry also gives an indication of the data quality for the positions used to infer the phylogenetic classification. The column indicating the quality will contain the label _good_ if all phylogenetic positions contained in the set are covered at least 10fold and show a frequency of at least 75%, _bad_ if any phylogenetic position does not meet these, and _ugly_ if any phylogenetic position does not have a clear base call.                                                                                                                                                                                                                       | `TBstrains`         | `Classification/Strain_Classification.tab`                                                                                     |
| Mapping and variant statistics | For calculating an overview of mapping quality and detected variants for a dataset, using the SAMTOOLS flagstat program. The generated tab-delimited file stores all sample statistics for the analyzed datasets present in the working environment. This module employs the same thresholds to discern unambiguously covered positions as the TBvariants module (set by the --mincovf, --mincovr, --minphred20, --minfreq).                                                                                                                                                                                                                                                                                                                                                                                            | `TBstats`           | `Statistics/Mapping_and_Variant_Statistics.tab`                                                                                |
| SNP distance matrix            | For inferring likely related isolates based on pairwise distance of distinct SNP positions. In an agglomerative process, samples are grouped together if they are within the threshold set by the --distance OPTION. The output files consist of a text file listing the detected groups and ungrouped isolates, and the calculated pairwise distance matrix.                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `TBgroups`          | `Groups/[PROJECT]_joint_[mincovf]_[mincovr]_[minfreq]_[minphred20]_samples_amended_[unambig]_phylo_[window].matrix`            |
| Cluster groups                 | For inferring likely related isolates based on pairwise distance of distinct SNP positions. In an agglomerative process, samples are grouped together if they are within the threshold set by the --distance OPTION. The output files consist of a text file listing the detected groups and ungrouped isolates, and the calculated pairwise distance matrix.                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `TBgroups`          | `Groups/[PROJECT]_joint_[mincovf]_[mincovr]_[minfreq]_[minphred20]_samples_amended_[unambig]_phylo_[window]_[distance].groups` |
| Phylogenetic tree              | For post-processing of joint variant tables. This step produces a comprehensive variant table including calculated summary values for each position. In addition, the set of positions will be processed in consecutive filtering steps. In the first step, positions will be excluded if less than a minimum percentage of samples have unambiguous base calls at this position, with this threshold set by the --unambig OPTION. In addition, all samples need to have either a SNP or wild type base at the position, and positions within repetitive regions of the reference genome or within a resistance associated genes are excluded. The second filtering step removes positions that are located within a maximum distance to each other in the same sample, with this threshold set by the --window OPTION. | `TBamend`           | `Amend/[PROJECT]_joint_[mincovf]_[mincovr]_[minfreq]_[minphred20]_samples_amended_[unambig]_phylo.fasta`                       |
